PCbuild/regen.targets

```
@@ -76,10 +76,20 @@
     <Exec Command="$(PythonForBuild) Tools\build\generate_global_objects.py"
           WorkingDirectory="$(PySourcePath)" />
   </Target>
+  
+  <Target Name="_RegenJIT" Condition="'$(UseJIT)' == 'true'" DependsOnTargets="_UpdatePyconfig;FindPythonForBuild">
+    <PropertyGroup>
+      <JITArgs Condition="$(Platform) == 'ARM64'">aarch64-pc-windows-msvc</JITArgs>
+      <JITArgs Condition="$(Platform) == 'Win32'">i686-pc-windows-msvc</JITArgs>
+      <JITArgs Condition="$(Platform) == 'x64'">x86_64-pc-windows-msvc</JITArgs>
+      <JITArgs Condition="$(Configuration) == 'Debug'">$(JITArgs) --debug</JITArgs>
+    </PropertyGroup>
+    <Exec Command='$(PythonForBuild) "$(PySourcePath)Tools\jit\build.py" $(JITArgs)' WorkingDirectory="$(IntDir)"/>
+  </Target>
 
   <Target Name="Regen"
           Condition="$(Configuration) != 'PGUpdate'"
-          DependsOnTargets="_TouchRegenSources;_RegenPegen;_RegenAST_H;_RegenTokens;_RegenKeywords;_RegenGlobalObjects">
+          DependsOnTargets="_TouchRegenSources;_RegenPegen;_RegenAST_H;_RegenTokens;_RegenKeywords;_RegenGlobalObjects;_RegenJIT">
```

@zooba, the PGO issue is with this `Condition`. We don't regen during the `PGUpdate` step, and since `jit_stencils.h` lives in the build directory, I think it actually needs to be rebuilt there (unlike the rest of these files, which live in the source tree and are version-controlled).

This seems to fix it. What do you think?

```suggestion
  <Target Name="_RegenNoPGUpdate"
          Condition="$(Configuration) != 'PGUpdate'"
          DependsOnTargets="_TouchRegenSources;_RegenPegen;_RegenAST_H;_RegenTokens;_RegenKeywords;_RegenGlobalObjects">
  </Target>

  <Target Name="Regen" DependsOnTargets="_RegenNoPGUpdate;_RegenJIT">
```

---

How about we store it in `$(GeneratedPyConfigDir)` (defined in `pyproject.props`) instead of `$(IntDir)`? We hit a similar issue with the build-time generated pyconfig.h, and fixing its location in the build directory was the answer (we did the same with frozen modules, but as those are shared for all architectures it's probably no good here).

---

> How about we store it in `$(GeneratedPyConfigDir)` (defined in `pyproject.props`) instead of `$(IntDir)`?

Just changing `WorkingDirectory="$(IntDir)"` to `WorkingDirectory="$(GeneratedPyConfigDir)"` for the JIT build command doesn't fix the PGO problem for me locally. Am I missing a step?

---

Hm, it appears that `GeneratedPyConfigDir` has `Configuration` (`PGInstrument`, `PGUpdate`, etc.) as part of its path. So doesn't really fix the issue, unfortunately.

Should I just go with what I had in my comment above?

---

> it appears that `GeneratedPyConfigDir` has `Configuration` (`PGInstrument`, `PGUpdate`, etc.) as part of its path.

Hmm... yeah, it probably shouldn't. The only thing that changes in that file is the nogil flag. I wonder if this is why I was seeing more rebuilding happening during PGUpdate than I expected?

Since you're testing both, could you try removing `Configuration` from its definition? We really do not want to change any source files between PGInstrument and PGUpdate - the second stage should be purely linking.

---

> Hmm... yeah, it probably shouldn't. The only thing that changes in that file is the nogil flag.

Well, I think it inherits that part from `IntDir`, which is used by more than just `pyconfig.h`, right?

> Since you're testing both, could you try removing `Configuration` from its definition?

It "works", but I don't think it's correct in our case: we can't use the same `jit_stencils.h` file for release and debug builds. :(

I've gotten the incremental build working though, so at least I'm making *some* progress.

(Is there anything wrong with what I suggested above in my first comment, or are we just exploring different solutions?)

---

> Well, I think it inherits that part from `IntDir`

The flag is in both, but because we redistribute the header it also gets modified to contain the flag explicitly. IIRC, the issue was that other parts of our build end up with different `IntDir` but needed to be able to include the header - that's not the case here, we only need to import it in the same project that generates it.

> It "works", but I don't think it's correct in our case: we can't use the same `jit_stencils.h` file for release and debug builds. :(

You can't generate the both sets of code with `#ifdef` in it? If not, then maybe a switch on `Debug` and `!= Debug` to choose the directory using a new variable is the way to go.

Aren't we supposed to be checking this file into the repo anyway? Possibly we're getting very sidetracked by the fact that it can run during build, when we should just be running the regen script independently and most people will use the generated file.

> Is there anything wrong with what I suggested above in my first comment

Rerunning generation between PGO profiling and collecting? Yeah, in theory there is, though in practice we won't get burnt by it until somebody affects what the script produces. That seems likely to happen in the future, so let's just dodge it by doing it right now.

---

> You can't generate the both sets of code with #ifdef in it? If not, then maybe a switch on `Debug` and `!= Debug` to choose the directory using a new variable is the way to go.
> 
> Aren't we supposed to be checking this file into the repo anyway? Possibly we're getting very sidetracked by the fact that it can run during build, when we should just be running the regen script independently and most people will use the generated file.

This file isn't going to be tracked, as it's very platform-specific (and even specific to what's in your current `pyconfig.h`). The issue isn't just debug vs. non-debug, it's every configuration that changes what code is generated... including whether a build is free-threaded, etc.

Backing up: what I *need* is a way to generate a header file (after `pyconfig.h` but before the actual interpreter is compiled), that is basically super-specific to that particular build. I just copied-and-pasted from the `pyconfig.h` generation process since that seemed similar, but it's becoming clear that it's not similar enough.

What's the easiest way to do that, at least initially?

---

(Also, assume I don't really know how `PCbuild` works now, or why things are done the way they are.)

---

I think we've got probably the easiest way. If you're relying on reading `pyconfig.h` to generate it, you'll need to add a `DependsOnTargets` entry to your new target to make sure the target that generates `pyconfig.h` runs, and probably pass in its directory to your script as it isn't in an easily predictable location at this stage. If you don't, you'll have to pass in the `DisableGil` option explicitly, as there's no other way to tell.

The main issue is that we've never had to generate pyconfig on Windows before, so none of the infrastructure is built around that (otherwise there'd be a two step process like `configure`, probably). Nogil has changed that, but rather than forcing everyone to change their build process we're hacking it in (in the hope that in a couple years time we go back to a static config). It's honestly better to treat that as part of the platform than as a normal config option, but I don't know if that makes sense for your script.

And if this is the script that requires installing llvm, we should _definitely_ pregenerate the scripts. At worst, we generate and host them somewhere and download them on demand, but we shouldn't be making people install an entire second compiler toolset for this.

---

Okay, I'll stick with what this branch has now (except that I'll depend on and use the generated `pyconfig.h` directory, as you suggest).

> And if this is the script that requires installing llvm, we should *definitely* pregenerate the scripts. At worst, we generate and host them somewhere and download them on demand, but we shouldn't be making people install an entire second compiler toolset for this.

It is indeed the step that requires LLVM, but I can open another issue to discuss this further (I don't think it should block this particular PR). I agree that it's not ideal, but we should also realize that checking in or hosting the generated file this basically amounts to checking in or hosting compiled binaries. So it's really tricky to get right and there's a lot of subtlety involved (and if we go that route, we should try to find a solution that works for lots of platforms, not just `PCbuild`).

---

It could be a big ask for CI systems and PR builds, though. As long as we don't enable PR checks for the experiment, I'll be okay with requiring regeneration during build. If that's part of the plan though, then we should probably make a branch in cpython-bin-deps and put the prebuilt "binaries" in there, as we do for other prebuild dependencies.

But that can come later if the checks aren't on all the time. (Running checks only when JIT-specific files are modified would be okay, as we do for MSI changes on Windows.)

---

> It could be a big ask for CI systems and PR builds, though. As long as we don't enable PR checks for the experiment, I'll be okay with requiring regeneration during build. If that's part of the plan though, then we should probably make a branch in cpython-bin-deps and put the prebuilt "binaries" in there, as we do for other prebuild dependencies.
>
> But that can come later if the checks aren't on all the time. (Running checks only when JIT-specific files are modified would be okay, as we do for MSI changes on Windows.)

Yup, all of the JIT builds are only configured to run in CI when we touch files matching the pattern `**jit**`.

I'll tag you on a new issue once this merges and we can explore ways to make JIT builds more ergonomic for when things aren't so experimental anymore (I know @vstinner had concerns at the sprint, too).